### DISCUSSION TITLE
RuntimeError: Sizes of tensors must match except in dimension 4. Got 4 and 3 (The offending index is 0)

### DISCUSSION THREAD

**Discussion by amine0110** [Project-MONAI/MONAI#2068] [Q&A]
**Describe the bug**
Hi there
I am trying to do 3D segmentation from the spleen example. I took the program as it is, but I divided each **nifti file into group of 30 slices instead of 300** at least (I am working on full body tumor segmentation).
So I have got this problem:
`RuntimeError: Sizes of tensors must match except in dimension 4. Got 4 and 3 (The offending index is 0)`

I am trying to solve this problem since a week but I don't know where is the problem, I checked all the input lengths it is the same but I don't know.

The problem is on this line:
`outputs = model(inputs)`

And this is also with the output error:
`    127         if self.mode == "cat":`
`--> 128             return torch.cat([x, y], dim=self.dim)`
`    129         if self.mode == "add":`
`    130             return torch.add(x, y)`

Can you tell me where is the problem please.
Thank you.
**To Reproduce**
2. Install '....'
3. Run commands '....'

**Expected behavior**

**Screenshots**

**Environment**

Ensuring you use the relevant python executable, please paste the output of:

```
================================
Printing MONAI config...
================================
MONAI version: 0.6.dev2116
Numpy version: 1.19.5
Pytorch version: 1.8.1+cu101
MONAI flags: HAS_EXT = False, USE_COMPILED = False
MONAI rev id: c6ed5fc811d451554b28d63c773c08695f5e0a2a

Optional dependencies:
Pytorch Ignite version: NOT INSTALLED or UNKNOWN VERSION.
Nibabel version: 3.0.2
scikit-image version: 0.16.2
Pillow version: 7.1.2
Tensorboard version: 2.4.1
gdown version: 3.6.4
TorchVision version: 0.9.1+cu101
ITK version: NOT INSTALLED or UNKNOWN VERSION.
tqdm version: 4.60.0
lmdb version: 0.99
psutil version: 5.4.8

For details about installing the optional dependencies, please visit:
    https://docs.monai.io/en/latest/installation.html#installing-the-recommended-dependencies


================================
Printing system config...
================================
System: Linux
Linux version: Ubuntu 18.04.5 LTS
Platform: Linux-4.19.112+-x86_64-with-Ubuntu-18.04-bionic
Processor: x86_64
Machine: x86_64
Python version: 3.7.10
Process name: python3
Command: ['/usr/bin/python3', '-m', 'ipykernel_launcher', '-f', '/root/.local/share/jupyter/runtime/kernel-8287949f-9c85-4e88-bd06-417fe8553e58.json']
Open files: [popenfile(path='/usr/local/lib/python3.7/dist-packages/matplotlib/mpl-data/fonts/ttf/DejaVuSans.ttf', fd=69, position=69632, mode='r', flags=557056), popenfile(path='/usr/local/lib/python3.7/dist-packages/matplotlib/mpl-data/fonts/ttf/DejaVuSans.ttf', fd=70, position=69632, mode='r', flags=557056)]
Num physical CPUs: 1
Num logical CPUs: 2
Num usable CPUs: 2
CPU usage (%): [4.7, 5.4]
CPU freq. (MHz): UNKNOWN for given OS
Load avg. in last 1, 5, 15 mins (%): UNKNOWN for given OS
Disk usage (%): 57.2
Avg. sensor temp. (Celsius): UNKNOWN for given OS
Total physical memory (GB): 12.7
Available memory (GB): 10.4
Used memory (GB): 3.8

================================
Printing GPU config...
================================
Num GPUs: 1
Has CUDA: True
CUDA version: 10.1
cuDNN enabled: True
cuDNN version: 7603
Current device: 0
Library compiled for CUDA architectures: ['sm_37', 'sm_50', 'sm_60', 'sm_70']
GPU 0 Name: Tesla T4
GPU 0 Is integrated: False
GPU 0 Is multi GPU board: False
GPU 0 Multi processor count: 40
GPU 0 Total memory (GB): 14.8
GPU 0 Cached memory (GB): 1.2
GPU 0 Allocated memory (GB): 0.2
GPU 0 CUDA capability (maj.min): 7.5
```

**Additional context**
I am running the program in google colab.

**Comment by wyli** [comment]
perhaps you could use a `EnsureChannelFirstd ` after the loadimage transform (and remove any `addchannel` transform) https://docs.monai.io/en/latest/transforms.html#ensurechannelfirstd

**Reply by amine0110** [reply]
I did that, but always the same problem !

`input data information of the runtime error transform:
input data information of the runtime error transform:
image statistics:
Type: <class 'numpy.ndarray'>
Shape: (1, 340, 314, 14)
Value range: (0.0, 1.0)
image statistics:
Type: <class 'numpy.ndarray'>
Shape: (1, 340, 314, 14)
Value range: (0.0, 1.0)
label statistics:
Type: <class 'numpy.ndarray'>
Shape: (1, 340, 314, 14)
Value range: (0.0, 0.0)
label statistics:
Type: <class 'numpy.ndarray'>
Shape: (1, 340, 314, 14)
Value range: (0.0, 0.0)
image_meta_dict statistics:
Type: <class 'dict'>
Value: {'sizeof_hdr': array(348, dtype=int32), 'extents': array(0, dtype=int32), 'session_error': array(0, dtype=int16), 'dim_info': array(0, dtype=uint8), 'dim': array([  3, 512, 512,  29,   1,   1,   1,   1], dtype=int16), 'intent_p1': array(0., dtype=float32), 'intent_p2': array(0., dtype=float32), 'intent_p3': array(0., dtype=float32), 'intent_code': array(0, dtype=int16), 'datatype': array(4, dtype=int16), 'bitpix': array(16, dtype=int16), 'slice_start': array(0, dtype=int16), 'pixdim': array([-1.      ,  0.785156,  0.785156,  0.9     ,  1.      ,  1.      ,
        1.      ,  1.      ], dtype=float32), 'vox_offset': array(0., dtype=float32), 'scl_slope': array(nan, dtype=float32), 'scl_inter': array(nan, dtype=float32), 'slice_end': array(0, dtype=int16), 'slice_code': array(0, dtype=uint8), 'xyzt_units': array(2, dtype=uint8), 'cal_max': array(0., dtype=float32), 'cal_min': array(0., dtype=float32), 'slice_duration': array(0., dtype=float32), 'toffset': array(0., dtype=float32), 'glmax': array(0, dtype=int32), 'glmin': array(0, dtype=int32), 'qform_code': array(0, dtype=int16), 'sform_code': array(2, dtype=int16), 'quatern_b': array(0., dtype=float32), 'quatern_c': array(1., dtype=float32), 'quatern_d': array(0., dtype=float32), 'qoffset_x': array(213.2, dtype=float32), 'qoffset_y': array(-215.61472, dtype=float32), 'qoffset_z': array(-265.4, dtype=float32), 'srow_x': array([ -0.785156,   0.      ,   0.      , 213.2     ], dtype=float32), 'srow_y': array([  -0.      ,    0.785156,    0.      , -215.61472 ], dtype=float32), 'srow_z': array([   0. ,   -0. ,    0.9, -265.4], dtype=float32), 'affine': array([[   1.        ,    0.        ,    0.        , -187.80000305],
       [   0.        ,    1.        ,    0.        , -215.61471558],
       [   0.        ,    0.        ,    2.        , -265.3999939 ],
       [   0.        ,    0.        ,    0.        ,    1.        ]]), 'original_affine': array([[  -0.78515601,    0.        ,    0.        ,  213.19999695],
       [  -0.        ,    0.78515601,    0.        , -215.61471558],
       [   0.        ,   -0.        ,    0.89999998, -265.3999939 ],
       [   0.        ,    0.        ,    0.        ,    1.        ]]), 'as_closest_canonical': False, 'spatial_shape': array([512, 512,  29], dtype=int16), 'original_channel_dim': 'no_channel', 'filename_or_obj': '/content/drive/MyDrive/Colab Notebooks/tap_segmentation/TrainData/patient37_IMG0378.nii.gz'}
image_meta_dict statistics:
Type: <class 'dict'>
Value: {'sizeof_hdr': array(348, dtype=int32), 'extents': array(0, dtype=int32), 'session_error': array(0, dtype=int16), 'dim_info': array(0, dtype=uint8), 'dim': array([  3, 512, 512,  29,   1,   1,   1,   1], dtype=int16), 'intent_p1': array(0., dtype=float32), 'intent_p2': array(0., dtype=float32), 'intent_p3': array(0., dtype=float32), 'intent_code': array(0, dtype=int16), 'datatype': array(4, dtype=int16), 'bitpix': array(16, dtype=int16), 'slice_start': array(0, dtype=int16), 'pixdim': array([-1.      ,  0.785156,  0.785156,  0.9     ,  1.      ,  1.      ,
        1.      ,  1.      ], dtype=float32), 'vox_offset': array(0., dtype=float32), 'scl_slope': array(nan, dtype=float32), 'scl_inter': array(nan, dtype=float32), 'slice_end': array(0, dtype=int16), 'slice_code': array(0, dtype=uint8), 'xyzt_units': array(2, dtype=uint8), 'cal_max': array(0., dtype=float32), 'cal_min': array(0., dtype=float32), 'slice_duration': array(0., dtype=float32), 'toffset': array(0., dtype=float32), 'glmax': array(0, dtype=int32), 'glmin': array(0, dtype=int32), 'qform_code': array(0, dtype=int16), 'sform_code': array(2, dtype=int16), 'quatern_b': array(0., dtype=float32), 'quatern_c': array(1., dtype=float32), 'quatern_d': array(0., dtype=float32), 'qoffset_x': array(213.2, dtype=float32), 'qoffset_y': array(-215.61472, dtype=float32), 'qoffset_z': array(-265.4, dtype=float32), 'srow_x': array([ -0.785156,   0.      ,   0.      , 213.2     ], dtype=float32), 'srow_y': array([  -0.      ,    0.785156,    0.      , -215.61472 ], dtype=float32), 'srow_z': array([   0. ,   -0. ,    0.9, -265.4], dtype=float32), 'affine': array([[   1.        ,    0.        ,    0.        , -187.80000305],
       [   0.        ,    1.        ,    0.        , -215.61471558],
       [   0.        ,    0.        ,    2.        , -265.3999939 ],
       [   0.        ,    0.        ,    0.        ,    1.        ]]), 'original_affine': array([[  -0.78515601,    0.        ,    0.        ,  213.19999695],
       [  -0.        ,    0.78515601,    0.        , -215.61471558],
       [   0.        ,   -0.        ,    0.89999998, -265.3999939 ],
       [   0.        ,    0.        ,    0.        ,    1.        ]]), 'as_closest_canonical': False, 'spatial_shape': array([512, 512,  29], dtype=int16), 'original_channel_dim': 'no_channel', 'filename_or_obj': '/content/drive/MyDrive/Colab Notebooks/tap_segmentation/TrainData/patient37_IMG0378.nii.gz'}
label_meta_dict statistics:
Type: <class 'dict'>
Value: {'sizeof_hdr': array(348, dtype=int32), 'extents': array(0, dtype=int32), 'session_error': array(0, dtype=int16), 'dim_info': array(0, dtype=uint8), 'dim': array([  3, 512, 512,  29,   1,   1,   1,   1], dtype=int16), 'intent_p1': array(0., dtype=float32), 'intent_p2': array(0., dtype=float32), 'intent_p3': array(0., dtype=float32), 'intent_code': array(0, dtype=int16), 'datatype': array(2, dtype=int16), 'bitpix': array(8, dtype=int16), 'slice_start': array(0, dtype=int16), 'pixdim': array([-1.      ,  0.785156,  0.785156,  0.9     ,  1.      ,  1.      ,
        1.      ,  1.      ], dtype=float32), 'vox_offset': array(0., dtype=float32), 'scl_slope': array(nan, dtype=float32), 'scl_inter': array(nan, dtype=float32), 'slice_end': array(0, dtype=int16), 'slice_code': array(0, dtype=uint8), 'xyzt_units': array(2, dtype=uint8), 'cal_max': array(0., dtype=float32), 'cal_min': array(0., dtype=float32), 'slice_duration': array(0., dtype=float32), 'toffset': array(0., dtype=float32), 'glmax': array(0, dtype=int32), 'glmin': array(0, dtype=int32), 'qform_code': array(0, dtype=int16), 'sform_code': array(2, dtype=int16), 'quatern_b': array(0., dtype=float32), 'quatern_c': array(1., dtype=float32), 'quatern_d': array(0., dtype=float32), 'qoffset_x': array(213.2, dtype=float32), 'qoffset_y': array(-215.61472, dtype=float32), 'qoffset_z': array(-239.3, dtype=float32), 'srow_x': array([ -0.785156,   0.      ,   0.      , 213.2     ], dtype=float32), 'srow_y': array([  -0.      ,    0.785156,    0.      , -215.61472 ], dtype=float32), 'srow_z': array([   0. ,   -0. ,    0.9, -239.3], dtype=float32), 'affine': array([[   1.        ,    0.        ,    0.        , -187.80000305],
       [   0.        ,    1.        ,    0.        , -215.61471558],
       [   0.        ,    0.        ,    2.        , -239.30000305],
       [   0.        ,    0.        ,    0.        ,    1.        ]]), 'original_affine': array([[  -0.78515601,    0.        ,    0.        ,  213.19999695],
       [  -0.        ,    0.78515601,    0.        , -215.61471558],
       [   0.        ,   -0.        ,    0.89999998, -239.30000305],
       [   0.        ,    0.        ,    0.        ,    1.        ]]), 'as_closest_canonical': False, 'spatial_shape': array([512, 512,  29], dtype=int16), 'original_channel_dim': 'no_channel', 'filename_or_obj': '/content/drive/MyDrive/Colab Notebooks/tap_segmentation/TrainLabels/patient37_IMG0407.nii.gz'}
label_meta_dict statistics:
Type: <class 'dict'>
Value: {'sizeof_hdr': array(348, dtype=int32), 'extents': array(0, dtype=int32), 'session_error': array(0, dtype=int16), 'dim_info': array(0, dtype=uint8), 'dim': array([  3, 512, 512,  29,   1,   1,   1,   1], dtype=int16), 'intent_p1': array(0., dtype=float32), 'intent_p2': array(0., dtype=float32), 'intent_p3': array(0., dtype=float32), 'intent_code': array(0, dtype=int16), 'datatype': array(2, dtype=int16), 'bitpix': array(8, dtype=int16), 'slice_start': array(0, dtype=int16), 'pixdim': array([-1.      ,  0.785156,  0.785156,  0.9     ,  1.      ,  1.      ,
        1.      ,  1.      ], dtype=float32), 'vox_offset': array(0., dtype=float32), 'scl_slope': array(nan, dtype=float32), 'scl_inter': array(nan, dtype=float32), 'slice_end': array(0, dtype=int16), 'slice_code': array(0, dtype=uint8), 'xyzt_units': array(2, dtype=uint8), 'cal_max': array(0., dtype=float32), 'cal_min': array(0., dtype=float32), 'slice_duration': array(0., dtype=float32), 'toffset': array(0., dtype=float32), 'glmax': array(0, dtype=int32), 'glmin': array(0, dtype=int32), 'qform_code': array(0, dtype=int16), 'sform_code': array(2, dtype=int16), 'quatern_b': array(0., dtype=float32), 'quatern_c': array(1., dtype=float32), 'quatern_d': array(0., dtype=float32), 'qoffset_x': array(213.2, dtype=float32), 'qoffset_y': array(-215.61472, dtype=float32), 'qoffset_z': array(-239.3, dtype=float32), 'srow_x': array([ -0.785156,   0.      ,   0.      , 213.2     ], dtype=float32), 'srow_y': array([  -0.      ,    0.785156,    0.      , -215.61472 ], dtype=float32), 'srow_z': array([   0. ,   -0. ,    0.9, -239.3], dtype=float32), 'affine': array([[   1.        ,    0.        ,    0.        , -187.80000305],
       [   0.        ,    1.        ,    0.        , -215.61471558],
       [   0.        ,    0.        ,    2.        , -239.30000305],
       [   0.        ,    0.        ,    0.        ,    1.        ]]), 'original_affine': array([[  -0.78515601,    0.        ,    0.        ,  213.19999695],
       [  -0.        ,    0.78515601,    0.        , -215.61471558],
       [   0.        ,   -0.        ,    0.89999998, -239.30000305],
       [   0.        ,    0.        ,    0.        ,    1.        ]]), 'as_closest_canonical': False, 'spatial_shape': array([512, 512,  29], dtype=int16), 'original_channel_dim': 'no_channel', 'filename_or_obj': '/content/drive/MyDrive/Colab Notebooks/tap_segmentation/TrainLabels/patient37_IMG0407.nii.gz'}
image_transforms statistics:
Type: <class 'list'>
Value: [{'class': 'Spacingd', 'id': 140366510962832, 'orig_size': (512, 512, 29), 'extra_info': {'meta_data_key': 'image_meta_dict', 'old_affine': array([[  -0.78515601,    0.        ,    0.        ,  213.19999695],
       [  -0.        ,    0.78515601,    0.        , -215.61471558],
       [   0.        ,   -0.        ,    0.89999998, -265.3999939 ],
       [   0.        ,    0.        ,    0.        ,    1.        ]]), 'mode': 'bilinear', 'padding_mode': 'border', 'align_corners': False}}, {'class': 'Orientationd', 'id': 140366511174544, 'orig_size': (402, 402, 14), 'extra_info': {'meta_data_key': 'image_meta_dict', 'old_affine': array([[  -1.        ,    0.        ,    0.        ,  213.19999695],
       [   0.        ,    1.        ,    0.        , -215.61471558],
       [   0.        ,    0.        ,    2.        , -265.3999939 ],
       [   0.        ,    0.        ,    0.        ,    1.        ]])}}, {'class': 'CropForegroundd', 'id': 140366510963344, 'orig_size': (402, 402, 14), 'extra_info': {'box_start': [35, 23, 0], 'box_end': [375, 337, 14]}}]
image_transforms statistics:
Type: <class 'list'>
Value: [{'class': 'Spacingd', 'id': 140366510962832, 'orig_size': (512, 512, 29), 'extra_info': {'meta_data_key': 'image_meta_dict', 'old_affine': array([[  -0.78515601,    0.        ,    0.        ,  213.19999695],
       [  -0.        ,    0.78515601,    0.        , -215.61471558],
       [   0.        ,   -0.        ,    0.89999998, -265.3999939 ],
       [   0.        ,    0.        ,    0.        ,    1.        ]]), 'mode': 'bilinear', 'padding_mode': 'border', 'align_corners': False}}, {'class': 'Orientationd', 'id': 140366511174544, 'orig_size': (402, 402, 14), 'extra_info': {'meta_data_key': 'image_meta_dict', 'old_affine': array([[  -1.        ,    0.        ,    0.        ,  213.19999695],
       [   0.        ,    1.        ,    0.        , -215.61471558],
       [   0.        ,    0.        ,    2.        , -265.3999939 ],
       [   0.        ,    0.        ,    0.        ,    1.        ]])}}, {'class': 'CropForegroundd', 'id': 140366510963344, 'orig_size': (402, 402, 14), 'extra_info': {'box_start': [35, 23, 0], 'box_end': [375, 337, 14]}}]
label_transforms statistics:
Type: <class 'list'>
Value: [{'class': 'Spacingd', 'id': 140366510962832, 'orig_size': (512, 512, 29), 'extra_info': {'meta_data_key': 'label_meta_dict', 'old_affine': array([[  -0.78515601,    0.        ,    0.        ,  213.19999695],
       [  -0.        ,    0.78515601,    0.        , -215.61471558],
       [   0.        ,   -0.        ,    0.89999998, -239.30000305],
       [   0.        ,    0.        ,    0.        ,    1.        ]]), 'mode': 'nearest', 'padding_mode': 'border', 'align_corners': False}}, {'class': 'Orientationd', 'id': 140366511174544, 'orig_size': (402, 402, 14), 'extra_info': {'meta_data_key': 'label_meta_dict', 'old_affine': array([[  -1.        ,    0.        ,    0.        ,  213.19999695],
       [   0.        ,    1.        ,    0.        , -215.61471558],
       [   0.        ,    0.        ,    2.        , -239.30000305],
       [   0.        ,    0.        ,    0.        ,    1.        ]])}}, {'class': 'CropForegroundd', 'id': 140366510963344, 'orig_size': (402, 402, 14), 'extra_info': {'box_start': [35, 23, 0], 'box_end': [375, 337, 14]}}]
label_transforms statistics:
Type: <class 'list'>
Value: [{'class': 'Spacingd', 'id': 140366510962832, 'orig_size': (512, 512, 29), 'extra_info': {'meta_data_key': 'label_meta_dict', 'old_affine': array([[  -0.78515601,    0.        ,    0.        ,  213.19999695],
       [  -0.        ,    0.78515601,    0.        , -215.61471558],
       [   0.        ,   -0.        ,    0.89999998, -239.30000305],
       [   0.        ,    0.        ,    0.        ,    1.        ]]), 'mode': 'nearest', 'padding_mode': 'border', 'align_corners': False}}, {'class': 'Orientationd', 'id': 140366511174544, 'orig_size': (402, 402, 14), 'extra_info': {'meta_data_key': 'label_meta_dict', 'old_affine': array([[  -1.        ,    0.        ,    0.        ,  213.19999695],
       [   0.        ,    1.        ,    0.        , -215.61471558],
       [   0.        ,    0.        ,    2.        , -239.30000305],
       [   0.        ,    0.        ,    0.        ,    1.        ]])}}, {'class': 'CropForegroundd', 'id': 140366510963344, 'orig_size': (402, 402, 14), 'extra_info': {'box_start': [35, 23, 0], 'box_end': [375, 337, 14]}}]
foreground_start_coord statistics:
Type: <class 'numpy.ndarray'>
Shape: (3,)
Value range: (0, 35)
foreground_start_coord statistics:
Type: <class 'numpy.ndarray'>
Shape: (3,)
Value range: (0, 35)
foreground_end_coord statistics:
Type: <class 'numpy.ndarray'>
Shape: (3,)
Value range: (14, 375)
foreground_end_coord statistics:
Type: <class 'numpy.ndarray'>
Shape: (3,)
Value range: (14, 375)
/usr/local/lib/python3.7/dist-packages/monai/transforms/utils.py:365: UserWarning: N foreground 0, N  background 986355,unable to generate class balanced samples.
  f"N foreground {len(fg_indices)}, N  background {len(bg_indices)},"
/usr/local/lib/python3.7/dist-packages/monai/transforms/utils.py:365: UserWarning: N foreground 0, N  background 1274087,unable to generate class balanced samples.
  f"N foreground {len(fg_indices)}, N  background {len(bg_indices)},"
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-48-d52e12d63090> in <module>()
     26         ##################
     27         print(len(inputs))
---> 28         outputs = model(inputs)
     29         print('batataaaaa')
     30         loss = loss_function(outputs, labels)

13 frames
/usr/local/lib/python3.7/dist-packages/monai/networks/layers/simplelayers.py in forward(self, x)
    126 
    127         if self.mode == "cat":
--> 128             return torch.cat([x, y], dim=self.dim)
    129         if self.mode == "add":
    130             return torch.add(x, y)

RuntimeError: Sizes of tensors must match except in dimension 3. Got 26 and 25 (The offending index is 0)`

This is the output

**Reply by LalehSeyyed** [reply]
Any progress on this error? I am suffering from a similar problem @amine0110

**Reply by amine0110** [reply]
Hi @LalehSeyyed ,
Yes, I fixed the problem, there are two things that you need to check:

- The values your are using for the **spacial_size** in the function `RandCropByPosNegLabeld` try to change them into bigger values (128 worked well for me).
- The second thing, is that you need to check how many slices you have before using that same last function (and check the depth of your pixels, because when you use the function `Spacingd` in the **pixdim** it will decrease the number of slices depending to the given values).

I hope that was useful.

**Reply by LalehSeyyed** [reply]
Thanks @amine0110 . Would you please expand the **pixdim** selection for Spacingd a little bit? My images are in nrrd format with 512, 512, z dimensions where z varies per patient. Then how can I choose pixdim?

**Reply by amine0110** [reply]
@LalehSeyyed 
pixdim, is to control the pixel dimensions (width, height and depth) it does not have any relation with the image dimensions !
